src/parser/ion_1_0/text/mod.rs

```diff
@@ -0,0 +1,2614 @@
+            return None;
+        }
+
+        let maybe_ivm = take_ivm(self.remaining);
+        match maybe_ivm {
+            Ok((remaining, ivm)) => {
+                self.remaining = remaining;
+                return match self.handle_ivm(ivm) {
+                    Ok(_) => self.next(),
+                    Err(e) => Some(Err(Err::Failure(IonError::from_format_error(
+                        self.remaining,
+                        FormatError::Text(e),
+                    )))),
+                };
+            }
+            Err(Err::Failure(e)) => return Some(Err(Err::Failure(e))),
+            _ => (),
+        }
+
+        let maybe_tlv =
+            preceded(eat_ws, take_top_level_value(self.current_table.clone()))(self.remaining);
+        match maybe_tlv {
+            Ok((remaining, (value, next))) => {
+                self.remaining = remaining;
+                return self.handle_meta_values(value, next);
+            }
+            Err(Err::Failure(e)) => return Some(Err(Err::Failure(e))),
+            _ => (),
+        };
+
+        let maybe_last = preceded(
+            eat_ws,
+            terminated(opt(take_value(self.current_table.clone())), eof),
+        )(self.remaining);
+        let last_err = match maybe_last {
+            Ok((remaining, value)) => {
+                self.remaining = remaining;
+                return if let Some(value) = value {
+                    self.handle_meta_values(value, None)
+                } else {
+                    self.next()
+                };
+            }
+            Err(Err::Failure(e)) => {
+                return Some(Err(Err::Failure(e)));
+            }
+            Err(e) => e,
+        };
+
+        if self.remaining.is_empty() {
+            None
+        } else {
+            Some(Err(last_err))
+        }
+    }
+}
+
+#[derive(Clone, Copy, Debug)]
+struct Ivm(u32, u32);
+
+/// System values are NOPs, currently triggered by a symbol which matches the IVM.
+fn is_system_value(value: &ion::Value) -> bool {
+    !value.has_annotations()
+        && value.value
+            == ion::Data::Symbol(Some(SymbolToken::Known {
+                text: "$ion_1_0".to_owned(),
+            }))
+}
+
+fn take_ivm(i: &str) -> IonResult<&str, Ivm> {
+    let (i, ivm) = preceded(
+        eat_ws,
+        terminated(
+            preceded(
+                tag("$ion_"),
+                separated_pair(
+                    take_while1(is_dec_digit),
+                    char('_'),
+                    take_while1(is_dec_digit),
+                ),
+            ),
+            nl,
+        ),
+    )(i)?;
+
+    let parse_version_number = |_: ParseIntError| -> nom::Err<IonError<&str>> {
+        Err::Failure(IonError::from_format_error(
+            i,
+            FormatError::Text(TextFormatError::TODO),
+        ))
+    };
+    let major = ivm.0.parse::<u32>().map_err(parse_version_number)?;
+    let minor = ivm.1.parse::<u32>().map_err(parse_version_number)?;
```

Comment:
The closure translates the parse error into an unrecoverable parse failure, which seems like it should be okay (though the error being named TODO doesn't help the situation here). I can replace it with matches for each version segment, though there is more code duplication in that case.

Reply:
Since the information of the error is getting dropped, what do you think of structuring it like this?

```rust
let (major, minor) = match (ivm.0.parse::<u32>(), ivm.1.parse::<u32>()) {
    (major: u32, minor: u32) => (major, minor),
    (_, _) => return Err::Failure(IonError::from_format_error(
            i,
            FormatError::Text(TextFormatError::TODO),
        )),
}
```

Reply:
Nice, I like that. Didn't even think of constructing a tuple there. Yeah, the error type doesn't have any additional information (yet: https://doc.rust-lang.org/std/num/struct.ParseIntError.html#method.kind) so it is currently fine to just drop it.